File: packages/honker/examples/notify_listen.py
"""Ephemeral pub/sub — pg_notify semantics on SQLite.

Starts a listener on the 'orders' channel in one asyncio task, fires
three notifications from another, and prints each received payload.
Wake latency is ~1 ms cross-process; same-process (this example) is
faster.

    python examples/notify_listen.py
"""

import asyncio
import os
import tempfile

import honker


async def consumer(db: honker.Database, received: list, stop: asyncio.Event):
    async for n in db.listen("orders"):
        received.append(n.payload)
        if len(received) == 3:
            stop.set()
            return


async def main():
    with tempfile.TemporaryDirectory() as d:
        db = honker.open(os.path.join(d, "app.db"))

        received: list = []
        stop = asyncio.Event()
        task = asyncio.create_task(consumer(db, received, stop))
        # Give the listener a moment to attach + snapshot MAX(id).
        await asyncio.sleep(0.05)

        for order_id in (1, 2, 3):
            with db.transaction() as tx:
                tx.notify("orders", {"id": order_id, "event": "placed"})

        await asyncio.wait_for(stop.wait(), timeout=5.0)
        task.cancel()
        try:
            await task
        except asyncio.CancelledError:
            pass

        print(f"received {len(received)} notifications:")
        for r in received:
            print(f"  {r}")
        assert len(received) == 3


if __name__ == "__main__":
    asyncio.run(main())

File: packages/honker-node/examples/notify_listen.js
// Ephemeral pub/sub — pg_notify semantics on SQLite.
//
// A subscriber sleeps on updateEvents().next() and wakes on every DB
// commit (same process or across processes on the same file), then
// reads new rows from _honker_notifications by id.
//
//   node examples/notify_listen.js

const lit = require('..');
const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');

const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'honker-'));
const db = lit.open(path.join(dir, 'app.db'));

async function run() {
  const received = [];
  const ev = db.updateEvents();

  // Snapshot the current high-water mark so we only see NEW
  // notifications.
  let lastSeen = db.query(
    "SELECT COALESCE(MAX(id), 0) AS m FROM _honker_notifications",
  )[0].m;

  const listener = (async () => {
    while (received.length < 3) {
      await ev.next();
      const rows = db.query(
        "SELECT id, channel, payload FROM _honker_notifications "
        + "WHERE channel='orders' AND id > ? ORDER BY id",
        [lastSeen],
      );
      for (const r of rows) {
        received.push(JSON.parse(r.payload));
        lastSeen = r.id;
      }
    }
  })();

  // Give the listener a tick to park on updateEvents.
  await new Promise((r) => setTimeout(r, 50));

  for (const id of [1, 2, 3]) {
    const tx = db.transaction();
    tx.notify('orders', { id, event: 'placed' });
    tx.commit();
  }

  await listener;
  ev.close();

  console.log(`received ${received.length} notifications:`);
  for (const p of received) console.log(`  ${JSON.stringify(p)}`);
  if (received.length !== 3)
    throw new Error(`expected 3, got ${received.length}`);
}

run()
  .catch((e) => { console.error(e); process.exitCode = 1; })
  .finally(() => fs.rmSync(dir, { recursive: true, force: true }));
